cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -121,6 +122,84 @@ Result<std::unique_ptr<KernelState>> CountInit(KernelContext*,
       static_cast<const CountOptions&>(*args.options));
 }
 
+// ----------------------------------------------------------------------
+// Distinct Count implementation
+
+template <typename Type>
+struct CountDistinctImpl : public ScalarAggregator {
+  using MemoTable = typename arrow::internal::HashTraits<Type>::MemoTableType;
+
+  explicit CountDistinctImpl(MemoryPool* memory_pool, CountOptions options)
+      : options(std::move(options)), memo_table_(new MemoTable(memory_pool, 0)) {}
+
+  Status Consume(KernelContext*, const ExecBatch& batch) override {
+    if (batch[0].is_array()) {
+      const ArrayData& arr = *batch[0].array();
+      auto visit_null = [&]() {
+        if (this->nulls > 0) return Status::OK();
+        ++this->nulls;
+        return Status::OK();
+      };
```

Comment:
IIUC, `this->nulls` is of interest in terms of it either being zero or non-zero, and if it is non-zero the actual value does not matters. The lambda has the check to try to short-circuit based on it being non-zero but I will argue that the conditional check can be more expensive than a simple integer add. So I would simplify it to an unconditional counter:
```c++
auto visit_null = [&]() {
  ++this->nulls;
  return Status::OK();
}
```

Reply:
Right. Basically you can have:
```c++
this->has_nulls = arr.GetNullCount() > 0;
auto visit_null = []() {};
auto visit_value = ...
```

Reply:
Yeah I was thinking about this as well. Would be great to have a visit call only for non nulls and if the array has null I would just add 1 to the output.
Thanks for the feedback, let me try to improve this part.

Reply:
Ok now I understand better, but still I was thinking about these lines:

1. The amount of code to make this "prettier" (using visitor, enable_if, hard-coded generators, etc) is going to result in more LoC and abstractions. Are there any reasons why we want to make this here? Perhaps I'm missing something.
2. Also, if we rely in implicit cast: Doesn't that make it worse? Because we will need to apply a casting every time the user wants to use a kernel. Explicit registration should result in faster kernel resolutions right?

Let me know what you think here please!
cc @edponce @pitrou @lidavidm